### DISCUSSION TITLE
Time-series Classification: Which transformers are suitable for use with panel data and `TimeSeriesForestClassifier`?

### DISCUSSION THREAD

**Discussion by jmwhyte** [alan-turing-institute/sktime#5767] [Q&A]
I have panel data, and am interested in TSC. As time is short, I'm interested in `TimeSeriesForestClassifier` as I need `feature_importances_`, which I can obtain directly. I've only seen examples of `TimeSeriesForestClassifier` used in a pipeline with `ColumnConcatenator`. I've tried this, but performance so far isn't great (not that close to the accuracy obtained using a tabular transformer with a sklearn classifier), although some of this might have to do with the limitations of grid searching available. (See this discussion: https://github.com/sktime/sktime/discussions/5735.) Does anyone have any suggestions on what I could try?

**Comment by fkiraly** [comment]
FYI to @YHallouard who is also working on applications which require TSF (or other) feature importances, and who has implemented the feature importance logic.

Other related issue is https://github.com/sktime/sktime/issues/4336, FYI @varelasaraiva, @pascalamadeus, which contains a `gist` on applying SHAP to general time series classifiers, so it might be a solution to using a different classifier of higher performance and still getting feature importances.

**Comment by jmwhyte** [comment]
I'll add some more background on my problem, as it may influence viable approaches given e.g. the recent comment of @pascalamadeus on  [#4336](https://github.com/sktime/sktime/issues/4336). 
My panel data has around 600 instances. For five timepoints, I have around 100 variables. It is not likely that any of these dimensions will increase substantially in the near term as the data is from an annual survey, and instances might have joined the survey only in recent times.

**Reply by fkiraly** [reply]
Hm, based on your description, my intuition would tell me to treat this problem as:

* non-temporal, or low temporal influence - you have more variables and very few time points
* factor analysis, possibly longitudinal - you say survey, and the dimensions imply low rank

So - and at the danger of being the hammer salesman telling you to get a screwdriver - maybe `sktime` is not your tool of choice, at least not until there are more transformers in the space of dimension reduction or factor analysis.

What I would do first is do a (mean centered) singular value decomposition on the matix of instances times items and time points (the latter two flattened). Then inspect loadings, singular values, elbow curve. I would also take the singular vectors and see whether there's any visible temporality to them.

For predictive modelling, I would use:

* scores after dimension reduction put in standard sklearn classifiers, like random forest, gradient boosting, linear regression
* cross-decomposition like [partial least squares](https://scikit-learn.org/stable/modules/generated/sklearn.cross_decomposition.PLSRegression.html), also available in `sklearn`

The above should all come with interpretability mode - feature importance, or classical statistical analysis (e.g., PLS).

**Reply by fkiraly** [reply]
... taking one step back, this gets us back to the "sklearn like map" on "which methods to use" and tells us sth about general "shapes" of the problem - interesting.

**Reply by jmwhyte** [reply]
Thanks for all of that @fkiraly. It looks as though I have a lot to learn. 
So I can get started on the track you describe, could you please provide a bit more explanation of:
"What I would do first is do a (mean centered) singular value decomposition **on the matix of instances times items and time points (the latter two flattened). "**
So, if there are m variables available for n time points, this is: 
1. a matrix with e.g. each row is an instance and the columns are, in order:
variable 0 from t=0,..,n, variable 1 from t=0,..,n, ..... variable m from t=0,..n ?
2. a sequence of matrices, one for each time t=0,...n, where the matrix at any time t has rows that are instances, and columns are the variables at time t? (Maybe this makes more sense if the intent is to have SVD done for each time point, and results are tracked over time, which seems to be what you were suggesting.)

For the "loadings, singular values, elbow curve" you describe, I'll see what I can find on that in sklearn or other python packages.

This is a pilot project, so there still might be other data sources and questions to explore if this is successful. I'm still grateful for all of the routines available in sktime, and the speedy responses to my queries. I hope I can play some kind of role in sktime at some point.

**Reply by fkiraly** [reply]
Here are some refereces, @jmwhyte:

* [sklearn PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA) - mathematically, this is equivalent to performing SVD, as you can also get the `components_`. The scores are obtained when running transform
* [sklearn factor analysis](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FactorAnalysis.html#sklearn.decomposition.FactorAnalysis) - this is one specific choice using Gaussian latents, "factor analysis" in general describes a range of similar methods

**Reply by fkiraly** [reply]
PCA will give you, if you constrain to a small rank `n_components`:

* an `n_components` times 500 (5 timepoints times 100 variables) matrix of components. Each row of 500 is a "typical direction of change". You can interpret these as "summary variables", or the dimensions along which the survey measures variation.
* a 600 times `n_components` matrix of scores. Each row can be interpreted as where the individual questionnaires lie in the coordinate system given by the components.

An example would be for `n_components=2`, you inspect the component vector and compare against items, and you think this might be qualitativiely interpreted as "product satisfaction" and "brand attachment", or sth like the [big five personality traits](https://en.wikipedia.org/wiki/Big_Five_personality_traits), depending on the items of course. You get this interpretation from the `components_` matrix.

You can then apply this interpretatio via the score matrix (obtained from `transform`), to see how the individual responses score on the axes.

**Reply by jmwhyte** [reply]
Much appreciated @fkiraly . I tried quickly today (on a matrix of instances as rows and all variables/their timepoints as columns) using `scipy.linalg.svd` followed by `sklearn.decomposition.TruncatedSVD` once I had an estimate of the number of components. (About 98% of variability explained by 3 components; I tried between 3 and 7 in the svd for a fitting trial.) However, using the transformation on the Xtest lead to scores (varying with a quick scan over `n_estimators` and `random_state` in `RandomForestClassifier` of between 0.48 (worse than the `DummyClassifier`) and about 0.6. This is much worse than what I'd managed with e.g. SummaryTransformer before the Random Forest. I suspected that this data had some non-linear relationships between features and labels, and that seems confirmed now. However, I'll still look at your links with interest in case I've missed something.

**Reply by fkiraly** [reply]
there may also be competing goals here - interpretability, or predictive accuracy.

A typical compromise when you use `PCA` is letting more `n_components` through.

**Reply by fkiraly** [reply]
> I suspected that this data had some non-linear relationships between features and labels, and that seems confirmed now.

Have you tried PLS?

**Reply by jmwhyte** [reply]
PLS is only a vague memory from many years ago. But as I'm keen on a new lead, I will give that a try.

**Reply by fkiraly** [reply]
mentioned here aleady,
https://github.com/sktime/sktime/discussions/5767#discussioncomment-8161319
PLS is just the abbreviation for partial least squares

**Reply by jmwhyte** [reply]
Regarding the suggestions of @fkiraly, I will add how I interpret these, and ask for advice if there's something I've misunderstood. Most seem straightforward once  you have the command, but one case is a little different...

"For predictive modelling, I would use:

scores after dimension reduction put in standard sklearn classifiers, like random forest, gradient boosting, linear regression

_dimension reduction can be achieved using `Sklearn.decomposition`, commands `PCA` or `FactorAnalysis`? It seems quite straightforward to transform an X matrix._

cross-decomposition like [partial least squares](https://scikit-learn.org/stable/modules/generated/sklearn.cross_decomposition.PLSRegression.html), also available in sklearn 

_here we define a PLSRegression instance, `plsr`, calculate `plsr.transform(Xtrain,ytrain)` and use the `plsr.x_scores_` in training the classifier `clf`? Similarly apply `plsr.predict(Xtest)` to get the x_scores for use in `clf.predict`? (That is, there is no need to keep any transformed y vector for later use?)_

PLSRegression has produced pretty good results with 6 components, with a best accuracy score found very quickly. I have slightly better accuracy without dimension reduction for my best method, but that took a lot of effort to tune. Clearly there's something further to explore here with a grid search to see if I can do any better.

**Reply by jmwhyte** [reply]
A quick note - PLS seems to do very well on my flattened data set. Having transformed the Xtrain through PLS, I can get a grid seach av score of around 0.97 with something like `GradientBoostClassifier`. Generalisation to the transformed Xtest has a score more like low-70ish percent. Making the upper end of `learning_rate` low in the grid search doesn't help much (which I hypothesised might be part of the problem - overfitting to training data). Using `RandomForestClassifier` doesn't do much better on the Xtest. Unfortunately, the strong preference for interpretability in this project means I will have to go back to some kind of random forest on the un-transformed data. Still, this has been a very educational experience, and the advice has been appreciated. If results are satisfactory here, it might smooth the way to further work.

**Comment by jmwhyte** [comment]
> ... taking one step back, this gets us back to the "sklearn like map" on "which methods to use" and tells us sth about general "shapes" of the problem - interesting.

Yes, I think the necessary knowledge might be in the community already. A relatively easy first step might even be a "traffic light" system: (red) sktime is probably not suitable for classification if your data is like this, consider transforming your data with sktime transformers for use with sklearn; (amber) sktime might be suitable if your data has this feature but not these features; (green) sktime is likely to be suitable for your application if the data is not in the red or amber categories.  Maybe others who know more than me could refine the general idea?

**Reply by fkiraly** [reply]
The crux is "suitable" - while you can stick certain data in, it does not always make sense.

My problem with categorical statements about "sktime is useful or not" is: you *could* implement longitudinal factor analysis as an `sktime` transformer, because longitudinal FA has the dimensions instance, item (variable), time. This would be most appropriate for your problem, and cannot be easily implemented in an `sklearn` interface, at least `sktime` is more fitting.

The problem is that `sktime` simply has a low amount of dimension reduction, factor analysis, and the like for time series panels. Might be a nice project for someone?

**Comment by YHallouard** [comment]
Hi @jmwhyte 

I understand that you have 600 windows of 100 multivariate time series, each window consisting of 5 data points. If this is correct, using TimeSeriesForestClassifier may not be as effective as using a traditional tabular classifier. As @fkiraly mentioned, extracting meaningful temporal information from such short windows is challenging.

TimeSeriesForestClassifier's strength lies in its ability to extract a multitude of features from randomly selected intervals within the data. However, with only 5 data points, the number of possible random intervals is limited, resulting in features that are essentially the same as those obtained from basic tabular analysis (e.g., mean, standard deviation, trend). Moreover, extracting only one or two features per 5 data points would be more efficient.

Another approach to dimensional reduction (from 5 data points to 1 feature) could involve excluding some predictors altogether. 100 predictors for 600 instances seems excessive, suggesting that some predictors may be redundant or irrelevant.

**Reply by jmwhyte** [reply]
I am investigating how to obtain more data, but I can see that what you write makes sense. Using sktime was very good for my only other TSC project, which had about 60 instances with 9 variables over 30 time points. So given my success and limited knowledge, I thought coming back to sktime would be a good way to get results in the short time allowed. There's no substitute for knowing the methods and their limitations. Still, sktime functions like `PCATransformer` and `SummaryTransformer` used with `RandomForestClassifier` have been pretty useful so far.

**Reply by fkiraly** [reply]
indeed, see above - based on the setting (survey) and the data dimensions, my intuition tells me SVD (PCA), factor analysis, or feature extraction, too.

**Reply by fkiraly** [reply]
> 100 predictors for 600 instances seems excessive, suggesting that some predictors may be redundant or irrelevant.

@YHallouard, this is typical in survey settings or psychometrics - the generic rationale is, your measurements are very noisy, so you put in more questions measuring "sort-of-the-same" thing.

**Reply by fkiraly** [reply]
as @YHallouard indicates, replacing the 5 longitudinal surveys with `mean` and `std` might also be interesting before you do SVD or factor analysis

**Comment by jmwhyte** [comment]
> as @YHallouard indicates, replacing the 5 longitudinal surveys with `mean` and `std` might also be interesting before you do SVD or factor analysis

Thanks, I'll add this to the list of possibilities for next week!

**Comment by jmwhyte** [comment]
> indeed, see above - based on the setting (survey) and the data dimensions, my intuition tells me SVD (PCA), factor analysis, or feature extraction, too.

When you talk of "feature extraction", do you suggest using some part of [sklearn's feature extraction](https://scikit-learn.org/stable/modules/feature_extraction.html) on the flattened matrix, or applying something like sktime's `TSFreshFeatureExtractor` to the panel data?

**Reply by fkiraly** [reply]
intuition-wise, sklearn, primarily, along the lines of (potentially sparse) factor analysis.
